// @cucumber/json-to-messages/dist/src/AstMaker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var messages_1 = require("@cucumber/messages");
var AstMaker = /** @class */ (function () {
    function AstMaker() {
    }
    AstMaker.prototype.makeGherkinDocument = function (uri, feature) {
        return messages_1.messages.GherkinDocument.create({
            uri: uri,
            feature: feature,
        });
    };
    AstMaker.prototype.makeFeature = function (line, keyword, name, description, children, tags) {
        return messages_1.messages.GherkinDocument.Feature.create({
            location: messages_1.messages.Location.create({ line: line }),
            keyword: keyword,
            name: name,
            description: description,
            children: children.map(function (child) { return child; }),
            tags: tags ? tags.map(function (tag) { return tag; }) : undefined,
        });
    };
    AstMaker.prototype.makeTag = function (name, line) {
        return messages_1.messages.GherkinDocument.Feature.Tag.create({
            name: name,
            location: messages_1.messages.Location.create({ line: line }),
        });
    };
    AstMaker.prototype.makeScenarioFeatureChild = function (id, line, keyword, name, description, steps, tags) {
        return messages_1.messages.GherkinDocument.Feature.FeatureChild.create({
            scenario: messages_1.messages.GherkinDocument.Feature.Scenario.create({
                id: id,
                location: messages_1.messages.Location.create({ line: line }),
                keyword: keyword,
                name: name,
                description: description,
                steps: steps.map(function (step) { return step; }),
                tags: tags ? tags.map(function (tag) { return tag; }) : undefined,
            }),
        });
    };
    AstMaker.prototype.makeBackgroundFeatureChild = function (line, keyword, name, description, steps) {
        return messages_1.messages.GherkinDocument.Feature.FeatureChild.create({
            background: messages_1.messages.GherkinDocument.Feature.Background.create({
                location: messages_1.messages.Location.create({ line: line }),
                keyword: keyword,
                name: name,
                description: description,
                steps: steps.map(function (step) { return step; }),
            }),
        });
    };
    AstMaker.prototype.makeStep = function (id, line, keyword, text, docString, dataTable) {
        return messages_1.messages.GherkinDocument.Feature.Step.create({
            id: id,
            location: messages_1.messages.Location.create({ line: line }),
            keyword: keyword,
            text: text,
            docString: docString,
            dataTable: dataTable,
        });
    };
    AstMaker.prototype.makeDocstring = function (mediaType, content) {
        return messages_1.messages.GherkinDocument.Feature.Step.DocString.create({
            mediaType: mediaType,
            content: content,
        });
    };
    AstMaker.prototype.makeDataTable = function (cells) {
        return messages_1.messages.GherkinDocument.Feature.Step.DataTable.create({
            rows: cells.map(function (row) {
                return messages_1.messages.GherkinDocument.Feature.TableRow.create({
                    cells: row.map(function (cell) {
                        return messages_1.messages.GherkinDocument.Feature.TableRow.TableCell.create({
                            value: cell,
                        });
                    }),
                });
            }),
        });
    };
    return AstMaker;
}());
exports.default = AstMaker;
