// @cucumber/json-to-messages/dist/src/PredictableHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var messages_1 = require("@cucumber/messages");
var SupportCodeExecutor_1 = require("./SupportCodeExecutor");
var PredictableHook = /** @class */ (function () {
    function PredictableHook(id, scenarioId, location, status, duration, errorMessage) {
        this.id = id;
        this.scenarioId = scenarioId;
        this.location = location;
        this.status = status;
        this.duration = duration;
        this.errorMessage = errorMessage;
    }
    PredictableHook.prototype.match = function (pickle) {
        if (!pickle.astNodeIds.includes(this.scenarioId)) {
            return null;
        }
        return new SupportCodeExecutor_1.NilCodeExecutor(null);
    };
    PredictableHook.prototype.toMessage = function () {
        var locationChunks = this.location.split(':');
        return messages_1.messages.Envelope.create({
            hook: messages_1.messages.Hook.create({
                id: this.id,
                sourceReference: messages_1.messages.SourceReference.create({
                    uri: locationChunks[0],
                    location: messages_1.messages.Location.create({
                        line: parseInt(locationChunks[1]),
                    }),
                }),
            }),
        });
    };
    return PredictableHook;
}());
exports.default = PredictableHook;
