// @cucumber/json-to-messages/dist/src/PredictableStepDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var messages_1 = require("@cucumber/messages");
var SupportCodeExecutor_1 = require("./SupportCodeExecutor");
var PredictableStepDefinition = /** @class */ (function () {
    function PredictableStepDefinition(id, stepId, location, status, duration, errorMessage) {
        this.id = id;
        this.stepId = stepId;
        this.location = location;
        this.status = status;
        this.duration = duration;
        this.errorMessage = errorMessage;
    }
    PredictableStepDefinition.prototype.match = function (pickleStep) {
        if (pickleStep.astNodeIds.includes(this.stepId)) {
            return new SupportCodeExecutor_1.NilCodeExecutor(this.id);
        }
        return null;
    };
    PredictableStepDefinition.prototype.toMessage = function () {
        var locationChunks = this.location.split(':');
        return messages_1.messages.Envelope.create({
            stepDefinition: messages_1.messages.StepDefinition.create({
                id: this.id,
                sourceReference: messages_1.messages.SourceReference.create({
                    uri: locationChunks[0],
                    location: messages_1.messages.Location.create({
                        line: parseInt(locationChunks[1]),
                    }),
                }),
            }),
        });
    };
    return PredictableStepDefinition;
}());
exports.default = PredictableStepDefinition;
