// @cucumber/json-to-messages/dist/src/PredictableSupportCode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var PredictableHook_1 = __importDefault(require("./PredictableHook"));
var messages_1 = require("@cucumber/messages");
var PredictableStepDefinition_1 = __importDefault(require("./PredictableStepDefinition"));
var PredictableSupportCode = /** @class */ (function () {
    function PredictableSupportCode(supportCode) {
        this.supportCode = supportCode;
    }
    PredictableSupportCode.prototype.addPredictableBeforeHook = function (location, scenarioId, status, duration, stack) {
        var id = this.supportCode.newId();
        this.supportCode.registerBeforeHook(new PredictableHook_1.default(id, scenarioId, location, this.statusFromString(status), duration || 0, stack));
    };
    PredictableSupportCode.prototype.addPredictableAfterHook = function (location, scenarioId, status, duration, stack) {
        var id = this.supportCode.newId();
        this.supportCode.registerAfterHook(new PredictableHook_1.default(id, scenarioId, location, this.statusFromString(status), duration || 0, stack));
    };
    PredictableSupportCode.prototype.addPredictableStepDefinition = function (location, stepId, status, duration, stack) {
        var id = this.supportCode.newId();
        this.supportCode.registerStepDefinition(new PredictableStepDefinition_1.default(id, stepId, location, this.statusFromString(status), duration || 0, stack));
    };
    PredictableSupportCode.prototype.statusFromString = function (status) {
        if (status === 'passed') {
            return messages_1.messages.TestStepFinished.TestStepResult.Status.PASSED;
        }
        if (status === 'pending') {
            return messages_1.messages.TestStepFinished.TestStepResult.Status.PENDING;
        }
        if (status === 'failed') {
            return messages_1.messages.TestStepFinished.TestStepResult.Status.FAILED;
        }
        return messages_1.messages.TestStepFinished.TestStepResult.Status.UNDEFINED;
    };
    return PredictableSupportCode;
}());
exports.default = PredictableSupportCode;
