// @cucumber/fake-cucumber/dist/src/dsl.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var messages_1 = require("@cucumber/messages");
var stack_utils_1 = __importDefault(require("stack-utils"));
var util_1 = require("util");
function setSupportCode(setSupportCode) {
    global.supportCode = setSupportCode;
}
exports.setSupportCode = setSupportCode;
function defineStepDefinition(expression, body) {
    global.supportCode.defineStepDefinition(getSourceReference(new Error().stack), expression, body);
}
function defineBeforeHook(tagExpressionOrBody, body) {
    global.supportCode.defineBeforeHook(getSourceReference(new Error().stack), tagExpressionOrBody, body);
}
function defineAfterHook(tagExpressionOrBody, body) {
    global.supportCode.defineAfterHook(getSourceReference(new Error().stack), tagExpressionOrBody, body);
}
function defineParameterType0(parameterTypeDefinition) {
    global.supportCode.defineParameterType(parameterTypeDefinition);
}
function getSourceReference(stackTrace) {
    var stack = new stack_utils_1.default({
        cwd: process.cwd(),
        internals: stack_utils_1.default.nodeInternals(),
    });
    var trace = stack.clean(stackTrace);
    var callSite = stack.parseLine(trace.split('\n')[1]);
    var uri = callSite.file, line = callSite.line;
    return new messages_1.messages.SourceReference({
        uri: uri,
        location: new messages_1.messages.Location({
            line: line,
        }),
    });
}
var Given = defineStepDefinition;
exports.Given = Given;
var When = defineStepDefinition;
exports.When = When;
var Then = defineStepDefinition;
exports.Then = Then;
var Before = defineBeforeHook;
exports.Before = Before;
var After = defineAfterHook;
exports.After = After;
var ParameterType = defineParameterType0;
exports.ParameterType = ParameterType;
var defineParameterType = util_1.deprecate(defineParameterType0, 'Please use ParameterType instead');
exports.defineParameterType = defineParameterType;
