// @cucumber/fake-cucumber/dist/src/PickleTestStep.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var TestStep_1 = __importDefault(require("./TestStep"));
var messages_1 = require("@cucumber/messages");
var PickleTestStep = /** @class */ (function (_super) {
    __extends(PickleTestStep, _super);
    function PickleTestStep() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PickleTestStep.prototype.toMessage = function () {
        return new messages_1.messages.TestCase.TestStep({
            id: this.id,
            pickleStepId: this.sourceId,
            stepDefinitionIds: this.supportCodeExecutors.map(function (supportCodeExecutor) { return supportCodeExecutor.stepDefinitionId; }),
            stepMatchArgumentsLists: this.supportCodeExecutors.map(function (e) {
                return new messages_1.messages.TestCase.TestStep.StepMatchArgumentsList({
                    stepMatchArguments: e.argsToMessages().slice(),
                });
            }),
        });
    };
    return PickleTestStep;
}(TestStep_1.default));
exports.default = PickleTestStep;
